app/src/heartbeat/heartbeat_storage_desktop.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIREBASE_APP_SRC_HEARTBEAT_HEARTBEAT_STORAGE_DESKTOP_H_
+#define FIREBASE_APP_SRC_HEARTBEAT_HEARTBEAT_STORAGE_DESKTOP_H_
+
+#include <ctime>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "app/logged_heartbeats_generated.h"
+
+namespace firebase {
+namespace heartbeat {
+
+using LoggedHeartbeatsFbs =
+    com::google::firebase::cpp::heartbeat::LoggedHeartbeats;
+
+struct LoggedHeartbeats {
+  // Last date for which a heartbeat was logged (YYYY-MM-DD).
+  std::string last_logged_date;
+
+  // Map from user agent to a list of dates (YYYY-MM-DD).
+  std::map<std::string, std::vector<std::string> > heartbeats;
+};
+
+// Reads and writes the last time heartbeat was sent for an SDK using persistent
+// storage.
+//
+// As this class uses the filesystem, there are several potential points of
+// failure:
+// - In the case of a failure, `Read` will return nullptr and `Write` will
+// return false;
+// - If a failure occurred, GetError can be used to get a descriptive error
+// message.
+// TODO(almostmatt): consider returning distinct error codes.
+class HeartbeatStorageDesktop {
+ public:
+  explicit HeartbeatStorageDesktop(std::string app_id);
+
+  // If the previous disk operation failed, contains additional details about
+  // the error; otherwise is empty.
+  const std::string& GetError() const { return error_; }
```

Comment:
I am not so funded to this pattern.

1. If changing this field is guarded by a mutex, you should guard it when getting this property as well, since this can be changed from a different thread. Even this property is guarded, when the called try to access it, the error may not be relevant to what the called did.
2. It is unclear from an API level that `ReadTo()` and `Write()` would change this property and expect the called to retrieve error using a different API. 

I can think of a couple of approaches for this
1. Change API to something like `bool ReadTo(LoggedHeartbeats* heartbeats_output, std::string* error);`. This makes `error` arg optional and should be handled in the implementation.
2. Just log the error.  This can be done in a couple of ways. Each FirebaseApp instance has its own logger. https://github.com/firebase/firebase-cpp-sdk/blob/main/app/src/app_common.cc#L186. You can keep a reference to `App`, or the logger itself. Or you can create your own logger for an instance of Heartbeat. But I think former is better since it is easier to set verbosity level.

Reply:
Ok. For now I removed the mutex.
Deciding whether to rely on some new mutex, or scheduler, or just OS file locking is something that I think can be added while working on the controller.

Reply:
Ok following your later suggestion, I added Logger* as a parameter in the constructor for the class, and anywhere that it previously set error_, it will now call logger_->LogError.

Reply:
Another pattern for error output that I have seen is to have Read/Write return
absl::StatusOr<Type>

Then if it succeeds the return value is Type, and if it fails the return value is a Status which can have an error message (like absl::InvalidArgumentError("bad mode"))
But I don't see absl::Status being used much in Firebase C++ SDK